File: jvm/src/main/scala/com/muchq/scraphics/tracer/Tracer.scala
package com.muchq.scraphics.tracer

import com.muchq.scraphics.tracer.Constants.{INF, EPSILON, BACKGROUND}

object Tracer {
  def drawScene(scene: Scene, image: Image, cameraPosition: Vec3): Unit =
    for
      x <- (-image.width / 2).toInt until (image.width / 2).toInt
      y <- (-image.height / 2).toInt until (image.height / 2).toInt
    do
      val direction: Vec3 = canvasToViewport(Vec2(x, y), image, scene)
      val color: Color = traceRay(cameraPosition, direction, 1.0, INF, scene, 2)
      image.putPixel(x, y, color)

  private def canvasToViewport(canvasPoint: Vec2, image: Image, scene: Scene): Vec3 =
    Vec3(canvasPoint._1 * scene.viewportSize / image.width, canvasPoint._2 * scene.viewportSize / image.height, scene.projectionPlane)

  private def intersectRaySphere(origin: Vec3, direction: Vec3, sphere: Sphere): (Double, Double) =
    val originToSphere: Vec3 = origin - sphere.center

    val a = direction dot direction
    val b = (originToSphere dot direction) * 2
    val c = (originToSphere dot originToSphere) - sphere.r2

    val discriminant: Double = b*b - (4 * a * c)
    if discriminant < 0 then
      return (INF, INF)

    val sqrtDiscr: Double = Math.sqrt(discriminant)
    val t1: Double = (-b + sqrtDiscr) / (2 * a)
    val t2: Double = (-b - sqrtDiscr) / (2 * a)
    (t1, t2)

  private def reflectRay(normal: Vec3, ray: Vec3): Vec3 =
    (normal * 2 * (normal dot ray)) - ray

  private def closestIntersection(origin: Vec3, direction: Vec3, tMin: Double, tMax: Double, spheres: List[Sphere]): (Option[Sphere], Double) =
    def intersectHelper(s: Sphere): (Option[Sphere], Double) =
      val (t1, t2) = intersectRaySphere(origin, direction, s)
      var closestSphere: Option[Sphere] = None
      var closestT: Double = INF
      if t1 < closestT && tMin < t1 && t1 < tMax then
        closestT = t1
        closestSphere = Some(s)

      if t2 < closestT && tMin < t2 && t2 < tMax then
        closestT = t2
        closestSphere = Some(s)
      (closestSphere, closestT)

    spheres.map(s => intersectHelper(s)).minBy(p => p._2)

  private def specularLightIntensity(normal: Vec3, ray: Vec3, view: Vec3, light: Light, specular: Double): Double =
    if specular <= 0 then
      return 0.0

    val reflectedRay: Vec3 = reflectRay(normal, ray)
    val rDotV: Double = reflectedRay dot view
    if rDotV > 0 then
      light.intensity * Math.pow(rDotV / reflectedRay.length() * view.length(), specular)
    else 0.0

  private def computeLighting(point: Vec3, normal: Vec3, view: Vec3, scene: Scene, specular: Double): Double =
    var intensity = 0.0
    for
      light <- scene.lights
    do
      if light.lightType == LightType.AMBIENT then
        intensity += light.intensity
      else
        val (ray, tMax) = light.lightType match
          case LightType.POINT => (light.position - point, 1.0)
          case LightType.DIRECTIONAL => (light.position, INF)
          case _ => ???

        val (shadowSphereMaybe, _) = closestIntersection(point, ray, EPSILON, tMax, scene.spheres)
        if shadowSphereMaybe.isEmpty then
          val nDotR = normal dot ray
          if nDotR > 0 then
            intensity += (light.intensity * (nDotR / (normal.length() * ray.length())))

          if specular > 0 then
            val reflectedRay = reflectRay(normal, ray)
            val rDotV = reflectedRay dot view
            if rDotV > 0 then
              intensity += (light.intensity * Math.pow(rDotV / (reflectedRay.length() * view.length()), specular))

    intensity

  private def traceRay(origin: Vec3, direction: Vec3, tMin: Double, tMax: Double, scene: Scene, recursionDepth: Int = 0): Color =
    val (closestSphereMaybe, closestT) = closestIntersection(origin, direction, tMin, tMax, scene.spheres)
    closestSphereMaybe match
      case None => BACKGROUND
      case Some(s) =>
        val point      = origin + (direction * closestT)
        val n          = point - s.center
        val normal     = n * (1 / n.length())

        val view       = direction * -1
        val lighting   = computeLighting(point, normal, view, scene, s.specular)
        val localColor = s.color * lighting

        if recursionDepth <= 0 || s.reflective <= 0 then
          return localColor

        val reflectedRay: Vec3 = reflectRay(normal, view)
        val reflectedColor: Color = traceRay(point, reflectedRay, EPSILON, INF, scene, recursionDepth - 1)
        localColor * (1.0 - s.reflective) + reflectedColor * s.reflective
}

File: domains/graphics/libs/tracy_go/tracer.go
package tracy

import "math"

func DrawScene(scene Scene, image *Image, camera Camera) {
	for x := -image.Width / 2; x < image.Width/2; x++ {
		for y := -image.Height / 2; y < image.Height/2; y++ {
			direction := camera.GetRayDirection(Vec2{float64(x), float64(y)}, image, scene)
			color := traceRay(camera.Position, direction, 1.0, Inf, scene, 2)
			image.PutPixel(float64(x), float64(y), color)
		}
	}
}

func canvasToViewport(canvasPoint Vec2, image *Image, scene Scene) Vec3 {
	return Vec3{
		X: canvasPoint.X * scene.ViewportSize / float64(image.Width),
		Y: canvasPoint.Y * scene.ViewportSize / float64(image.Height),
		Z: scene.ProjectionPlane,
	}
}

func intersectRaySphere(origin, direction Vec3, sphere Sphere) (float64, float64) {
	originToSphere := origin.Sub(sphere.Center)
	
	a := direction.Dot(direction)
	b := originToSphere.Dot(direction) * 2
	c := originToSphere.Dot(originToSphere) - sphere.R2
	
	discriminant := b*b - 4*a*c
	if discriminant < 0 {
		return Inf, Inf
	}
	
	sqrtDiscr := math.Sqrt(discriminant)
	t1 := (-b - sqrtDiscr) / (2 * a)
	t2 := (-b + sqrtDiscr) / (2 * a)
	return t1, t2
}

func reflectRay(normal, ray Vec3) Vec3 {
	return normal.Scale(2 * normal.Dot(ray)).Sub(ray)
}

func closestIntersection(origin, direction Vec3, tMin, tMax float64, spheres []Sphere) (*Sphere, float64) {
	var closestSphere *Sphere
	closestT := Inf
	
	for i := range spheres {
		t1, t2 := intersectRaySphere(origin, direction, spheres[i])
		
		if t1 < closestT && tMin < t1 && t1 < tMax {
			closestT = t1
			closestSphere = &spheres[i]
		}
		
		if t2 < closestT && tMin < t2 && t2 < tMax {
			closestT = t2
			closestSphere = &spheres[i]
		}
	}
	
	return closestSphere, closestT
}

func anyIntersection(origin, direction Vec3, tMin, tMax float64, spheres []Sphere) bool {
	// Optimization: precompute 'a' which is constant for the ray
	a := direction.Dot(direction)
	inv2a := 1.0 / (2 * a)

	for i := range spheres {
		// Inline intersection logic to avoid function call overhead and reuse 'a'
		sphere := spheres[i]
		originToSphere := origin.Sub(sphere.Center)

		b := originToSphere.Dot(direction) * 2
		c := originToSphere.Dot(originToSphere) - sphere.R2

		discriminant := b*b - 4*a*c
		if discriminant < 0 {
			continue
		}

		sqrtDiscr := math.Sqrt(discriminant)
		t1 := (-b - sqrtDiscr) * inv2a

		if tMin < t1 && t1 < tMax {
			return true
		}

		t2 := (-b + sqrtDiscr) * inv2a
		if tMin < t2 && t2 < tMax {
			return true
		}
	}
	return false
}

func computeLighting(point, normal, view Vec3, scene Scene, specular float64) float64 {
	intensity := 0.0
	
	for _, light := range scene.Lights {
		if light.Type == Ambient {
			intensity += light.Intensity
		} else {
			var ray Vec3
			var tMax float64
			
			switch light.Type {
			case Point:
				ray = light.Position.Sub(point)
				tMax = 1.0
			case Directional:
				ray = light.Position
				tMax = Inf
			}
			
			if !anyIntersection(point, ray, Epsilon, tMax, scene.Spheres) {
				nDotR := normal.Dot(ray)
				if nDotR > 0 {
					intensity += light.Intensity * (nDotR / (normal.Length() * ray.Length()))
				}
				
				if specular > 0 {
					reflectedRay := reflectRay(normal, ray)
					rDotV := reflectedRay.Dot(view)
					if rDotV > 0 {
						intensity += light.Intensity * math.Pow(rDotV/(reflectedRay.Length()*view.Length()), specular)
					}
				}
			}
		}
	}
	
	return intensity
}

func traceRay(origin, direction Vec3, tMin, tMax float64, scene Scene, recursionDepth int) Color {
	closestSphere, closestT := closestIntersection(origin, direction, tMin, tMax, scene.Spheres)
	
	if closestSphere == nil {
		return Background
	}
	
	point := origin.Add(direction.Scale(closestT))
	n := point.Sub(closestSphere.Center)
	normal := n.Scale(1 / n.Length())
	
	view := direction.Scale(-1)
	lighting := computeLighting(point, normal, view, scene, closestSphere.Specular)
	localColor := closestSphere.Color.Scale(lighting)
	
	if recursionDepth <= 0 || closestSphere.Reflective <= 0 {
		return localColor
	}
	
	reflectedRay := reflectRay(normal, view)
	reflectedColor := traceRay(point, reflectedRay, Epsilon, Inf, scene, recursionDepth-1)
	
	return localColor.Scale(1 - closestSphere.Reflective).Add(reflectedColor.Scale(closestSphere.Reflective))
}